### PR TITLE
Add gradients for add/sub/mul/div/sum/mean/transpose

### PR DESCRIPTION
**Pull request by dsmilkov**
Add gradients for `add`, `subtract`, `multiply`, `divide`, `sum`, `mean` and `transpose` with full support for broadcasting.

Also:
- Expose `Variable` to the public API.
- Remove usage of deprecated methods (e.g. `math.elementWiseMul`) in our own library

This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/544)

### CODE REVIEW

**Comment by caisq** [comment]
cc myself :)

**Comment by nkreeger** [comment]
Review status: 0 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/math/activation_functions.ts, line 39 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L2v2ECDZSzWy2EubZaW:-L2v2ECEY4JRpFs2R40v:bp9k3hw) ([raw file](https://github.com/pair-code/deeplearnjs/blob/238399f2b56dbdb260da8f47252699c741e31681/src/math/activation_functions.ts#L39)):*
> NDArray<'float32'>

Does casting here cause issues for ints/bools? Same in a couple of spots.

---

*[src/math/arithmetic_test.ts, line 173 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L2v4JbAAudmDeJUPp9C:-L2v4JbAAudmDeJUPp9D:b2fvdy5) ([raw file](https://github.com/pair-code/deeplearnjs/blob/238399f2b56dbdb260da8f47252699c741e31681/src/math/arithmetic_test.ts#L173)):*
> ('float32');
What about adding a couple of tests with int32 and bool?

---

*[src/math/arithmetic_test.ts, line 224 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L2v4TdafFAMNw4fEqyh:-L2v4TdafFAMNw4fEqyi:bktgzc0) ([raw file](https://github.com/pair-code/deeplearnjs/blob/238399f2b56dbdb260da8f47252699c741e31681/src/math/arithmetic_test.ts#L224)):*
> 'multiplyStrict
Ditto with multiply tests ('int32' and 'bool' expectations?).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544)*

**Comment by nkreeger** [comment]
:lgtm_strong:

---

Review status: 0 of 13 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-:-L2v51tATW5jB3l61egw:b-jbsf8z)*

**Comment by pyu10055** [comment]
Review status: 0 of 13 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/math/activation_functions.ts, line 38 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L2vPbYTeHiIckaNAI1S:-L2vPbYUfqI9ETCN4cE1:b-dba0i) ([raw file](https://github.com/pair-code/deeplearnjs/blob/238399f2b56dbdb260da8f47252699c741e31681/src/math/activation_functions.ts#L38)):*
> ```XML
>     return math.scope(() => {
>       const ySquared = math.multiplyStrict(y, y);
>       // 1 - y^2.
> ```

is derivative of tanh(x) 1 + tanh(x) ^ 2 ?

---

*[src/math/arithmetic_test.ts, line 117 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L2vTkVKHTsOTqvm45Cn:-L2vTkVL3tcvOi6VHX_3:b-rc2fhe) ([raw file](https://github.com/pair-code/deeplearnjs/blob/238399f2b56dbdb260da8f47252699c741e31681/src/math/arithmetic_test.ts#L117)):*
> ```XML
>       const dy = Scalar.new(4);
> 
>       const vjp = math.vjp(() => math.divide(a, b), {a, b}, dy);
> ```

is possible to use some more descriptive name for method like vjp and der?

---

*[src/math/arithmetic_test.ts, line 170 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L2vViIWO28JTjAtLI3M:-L2vViIWO28JTjAtLI3N:bgiwaeg) ([raw file](https://github.com/pair-code/deeplearnjs/blob/238399f2b56dbdb260da8f47252699c741e31681/src/math/arithmetic_test.ts#L170)):*
> ```XML
>       expect(vjp.a.shape).toEqual(a.shape);
>       expect(vjp.a.dtype).toEqual('float32');
>       test_util.expectArraysClose(vjp.a, [6 / 3 + 7 / 4 + 8 / 5]);
> ```

can you explain when broadcasting, why the derivative is sum of all broadcast vectors? thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544)*

**Comment by nsthorat** [comment]
:lgtm_strong:

---

Reviewed 7 of 13 files at r1, 2 of 5 files at r2, 1 of 1 files at r3.
Review status: 10 of 13 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[src/math/transpose_test.ts, line 63 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L33lI-29roN5fnJgrNS:-L33lI-29roN5fnJgrNT:b-jt858s) ([raw file](https://github.com/pair-code/deeplearnjs/blob/c982e67c919a7ae0210ccbb0fd05e62e97a04a4d/src/math/transpose_test.ts#L63)):*
> ```XML
>       const dy =
>           Array3D.new([2, 2, 2], [111, 211, 121, 221, 112, 212, 122, 222]);
>       const vjp = math.vjp(() => math.transpose(t, perm), {t}, dy);
> ```

you can just use "t" instead of "{t}" here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-:-L33nENcSK69U-zuZjEN:b-jbsf8z)*

**Comment by dsmilkov** [comment]
Reviewed 9 of 13 files at r1, 4 of 5 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[src/math/activation_functions.ts, line 38 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L2vPbYTeHiIckaNAI1S:-L33ba1eGUZmBIqiuvqS:bz01fim) ([raw file](https://github.com/pair-code/deeplearnjs/blob/238399f2b56dbdb260da8f47252699c741e31681/src/math/activation_functions.ts#L38)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

is derivative of tanh(x) 1 + tanh(x) ^ 2 ?
</blockquote></details>

it's `1 - tanh(x)^2`, which is `1 - y^2`

---

*[src/math/activation_functions.ts, line 39 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L2v2ECDZSzWy2EubZaW:-L33iri6H-5ooBGGpknb:bg1fcho) ([raw file](https://github.com/pair-code/deeplearnjs/blob/238399f2b56dbdb260da8f47252699c741e31681/src/math/activation_functions.ts#L39)):*
<details><summary><i>Previously, nkreeger (Nick Kreeger) wrote…</i></summary><blockquote>

> NDArray<'float32'>

Does casting here cause issues for ints/bools? Same in a couple of spots.
</blockquote></details>

not really worried, since we decided for gradients to always be floats. It makes no sense for gradients to ever be ints/bools, since the nature of the gradient is "what happens to y if I change x by tiny amount".

---

*[src/math/arithmetic_test.ts, line 117 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L2vTkVKHTsOTqvm45Cn:-L33jQawiRhaAJsV53wR:b-x44hc3) ([raw file](https://github.com/pair-code/deeplearnjs/blob/238399f2b56dbdb260da8f47252699c741e31681/src/math/arithmetic_test.ts#L117)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

is possible to use some more descriptive name for method like vjp and der?
</blockquote></details>

vjp stands for vector-jacobian product and it's pretty standard terminology in the auto-differentiation world. Note that users won't really have to use `vjp`, it's used mostly in testing.

TF eager calls it vjp too: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/eager/backprop.py#L619

---

*[src/math/arithmetic_test.ts, line 170 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L2vViIWO28JTjAtLI3M:-L33kl1uRJ71Tk-RU1UR:b-y17yfk) ([raw file](https://github.com/pair-code/deeplearnjs/blob/238399f2b56dbdb260da8f47252699c741e31681/src/math/arithmetic_test.ts#L170)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

can you explain when broadcasting, why the derivative is sum of all broadcast vectors? thanks
</blockquote></details>

when you broadcast you repeat the input multiple times so in the summation case:
[a] + [b, b, b] becomes [a, a, a] + [b1, b2, b3] = [y1, y2, y3]

then when you ask the question "how does changes in `a` affect changes in `y`, you realize the `a` affects both `y1`, `y2` and `y3`, at which point you have to sum those changes (standard chain rule), thus `da = dy1 + dy2 + dy3`

---

*[src/math/arithmetic_test.ts, line 173 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L2v4JbAAudmDeJUPp9C:-L33ldE6xVDiZZ6KouYI:b-7rl0jd) ([raw file](https://github.com/pair-code/deeplearnjs/blob/238399f2b56dbdb260da8f47252699c741e31681/src/math/arithmetic_test.ts#L173)):*
<details><summary><i>Previously, nkreeger (Nick Kreeger) wrote…</i></summary><blockquote>

> ('float32');
What about adding a couple of tests with int32 and bool?
</blockquote></details>

done.

---

*[src/math/arithmetic_test.ts, line 224 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L2v4TdafFAMNw4fEqyh:-L33pBPqWsm9RjwBj1L4:b6gg9d1) ([raw file](https://github.com/pair-code/deeplearnjs/blob/238399f2b56dbdb260da8f47252699c741e31681/src/math/arithmetic_test.ts#L224)):*
<details><summary><i>Previously, nkreeger (Nick Kreeger) wrote…</i></summary><blockquote>

> 'multiplyStrict
Ditto with multiply tests ('int32' and 'bool' expectations?).
</blockquote></details>

done. note that the gradient will be float32 regardless.

---

*[src/math/transpose_test.ts, line 63 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544#-L33lI-29roN5fnJgrNS:-L34DEv3n1pwH8ji6zH3:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/c982e67c919a7ae0210ccbb0fd05e62e97a04a4d/src/math/transpose_test.ts#L63)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

you can just use "t" instead of "{t}" here
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/544)*
